### ISSUE TITLE
obj.first caused Java::JavaUtil::NoSuchElementException on an empty Java21 ArrayList

### ISSUE THREAD

**Issue by jimtng** [jruby/jruby#8526]
Java21 introduced List.getFirst() and getLast(). So now this happens only in Java21 (jruby-9.4.9.0) because JRuby is calling the Java method. The same code runs fine and returns `nil` in Java17

```
irb(main):004:0> a = java.util.ArrayList.new
=> #<Java::JavaUtil::ArrayList: []>
irb(main):005:0> a
=> #<Java::JavaUtil::ArrayList: []>
irb(main):006:0> a.first
java.util.ArrayList.getFirst(java/util/ArrayList.java:439): Java::JavaUtil::NoSuchElementException
        from jdk.internal.reflect.DirectMethodHandleAccessor.invoke(jdk/internal/reflect/DirectMethodHandleAccessor.java:103)
        from java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:580)
        from org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:300)
        from org.jruby.javasupport.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:164)
        from RUBY.evaluate((irb):6)
        from org.jruby.RubyKernel.evalCommon(org/jruby/RubyKernel.java:1315)
        from org.jruby.RubyKernel.eval(org/jruby/RubyKernel.java:1277)
        from org.jruby.RubyKernel$INVOKER$s$0$3$eval.call(org/jruby/RubyKernel$INVOKER$s$0$3$eval.gen)
        from org.jruby.RubyKernel.loop(org/jruby/RubyKernel.java:1725)
        from org.jruby.RubyKernel$INVOKER$s$0$0$loop.call(org/jruby/RubyKernel$INVOKER$s$0$0$loop.gen)
        from org.jruby.RubyKernel.catch(org/jruby/RubyKernel.java:1426)
        from org.jruby.RubyKernel$INVOKER$s$rbCatch.call(org/jruby/RubyKernel$INVOKER$s$rbCatch.gen)
        from org.jruby.RubyKernel.catch(org/jruby/RubyKernel.java:1426)
        from org.jruby.RubyKernel$INVOKER$s$rbCatch.call(org/jruby/RubyKernel$INVOKER$s$rbCatch.gen)
        from RUBY.<main>(/home/openhab/.rbenv/versions/jruby-9.4.9.0/lib/ruby/gems/shared/gems/irb-1.4.2/exe/irb:11)
        from org.jruby.Ruby.runInterpreter(org/jruby/Ruby.java:1258)
        ... 13 levels...
```

**Comment by headius** [comment]
I wonder if there was some behavior change in Java 21??

**Comment by jimtng** [comment]
> I wonder if there was some behavior change in Java 21?

Are you referring to some interaction between java21 and jruby, or the fact that Java21 added getFirst() and getLast()?

JRuby needs to prioritise its own `#first` and `#last` over Java's `getFirst()` / `getLast()`. Where should I look?

**Comment by headius** [comment]
> the fact that Java21 added getFirst() and getLast()

Aha, that's the cause for certain.

JRuby's `first` and `last` won't ever raise an error, and will simply return `nil` if there's no elements. With the addition of Java 21's `getFirst` and `getLast`, those methods are getting overridden.

I tend to agree with you: list-like Java objects should behave like other list-like Ruby objects when called from Ruby, and the `first` and `last` methods should continue to work like Ruby users expect. The Java `getFirst` and `getLast` would remain in place to be called if that's the behavior you want.

**Comment by headius** [comment]
Aha, I knew we had dealt with this in the past.

This commit by @kares attempts to work around the issue by adding a `ruby_first` alias for our Ruby-like `first` logic: https://github.com/jruby/jruby/commit/7930401293c88b3bb0db2a36bfacd913bd064933

A workaround for you now would be to use `ruby_first` which won't be going anywhere

@kares I don't remember what discussions we had around this, but obviously at the time you or we decided not to prevent the Java `getFirst` from being bound as `first`. Maybe we can have a chat with @enebo about whether this is the best policy?

I really don't know the answer. I usually defer to whatever methods the Java class defines, but this is an alias we are giving it.  Might it be too frustrating to have Java collections that look like Ruby Arrays behave differently for just *some* of their methods.

**Comment by headius** [comment]
An alternative workaround is to open up the actual class for the objects that are erroring and re-define `first` to the Ruby behavior:

```ruby
class java::util::ArrayList
  alias first ruby_first # option 1
  def first = self[0] # option 2
end
```

We are debating whether the Ruby enhancement or the Java alias should take priority, and trying to figure out the best way to do the former.

**Comment by headius** [comment]
This would be a non-trivial fix to try to cram into 9.4.10 so we'll see about .11. Hoping to get some input from @kares on this one.

**Comment by ccutrer** [comment]
Might I point something out:

The Java method is `getFirst`. I can understand the reasoning for not wanting to hide built in Java methods with Ruby versions. And for `merge` there's no difference. But `first` is already a Ruby-only convenience alias for a Java method, so any callers are already saying "I prefer how Ruby does things". And `getFirst` is still available.

**Comment by headius** [comment]
@kares @ccutrer Yeah I think this is slightly different than the historical cases where the Java method *directly* conflicted with a Ruby convenience method. In this case, we're adding a Ruby alias from `getFirst` to `first` even though that method name is not defined in Ruby and we have a Ruby-friendly `first` on `List` that we'd actually like to call.

It's a non-trivial change because currently, the Java method binding logic only considers other Java methods (e.g. if a `first` did exist in Java, we don't want to overwrite it) or a global table of names it should not bind (we can't use that because it would stop us binding `first` in other places). So there needs to be some new logic to indicate that the Ruby-friendly alias of a Java method should be superseded by a utility method we already added to a superclass. Doable, but tricky.

**Comment by jimtng** [comment]
But Jruby #first and #last were/are actual ruby methods, even before java21. 

So should it not be about excluding certain java methods and more about selecting which method gets invoked in the ancestor chain?

**Comment by jimtng** [comment]
Furthermore, `java.util.SortedSet` has had `first()` and `last()` methods since at least [Java 8](https://docs.oracle.com/javase/8/docs/api/java/util/SortedSet.html).

In Java21, SortedSet implements java.util.SequencedCollection's `getFirst()` by [calling its first() method](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/SortedSet.html#getFirst()). So at least for SortedSet, we might want to call Java's version of `first()` (because it actually exists), but not otherwise. 

https://openjdk.org/jeps/431

**Comment by kares** [comment]
Having getFirst and first behave differently in this context (in my experience folks prefer the non-getter shorthand JRuby adds) is not a good idea and will just lead to more confusion, in general when upgrading Java versions. JRuby already has the ruby_ prefixed methods to solve these kind of issues and I would just support that here.

**Comment by jimtng** [comment]
JRuby makes Java Lists work just like Ruby's arrays and that's part of the beauty of JRuby.

Essentially we're choosing between having to special case calling Java's getFirst, vs having to special case calling Ruby's #first. 

By preferring to use Java's #first implementation, we'd have to always check whether the given array is Java's or Ruby's and branch the call to either #ruby_first, or #first, because [].ruby_first doesn't exist.

Or resort to a [more complicated code](https://github.com/openhab/openhab-jruby/commit/f5c1bd30c7ff5cef5a5da7acc2ec5de484a5a73e) just to avoid the exception.

**Comment by kares** [comment]
Java objects are Java first, in case it wasn't clear that is the lesson we learned with Java 8 leading to a bunch of conflicts and a lot of user confusion. Any deviation from that will IMHO lead to more confusion (again).

Your issue sounds like a problem of wanting to treat different objects the same way, which sounds like a bad design to start with. Maybe we can help your specific use-case if we understand more (what kind of types of collections are you dealing with)...

**Comment by headius** [comment]
@kares I agree that Java objects should be "Java first" but in this case we are explicitly adding method aliases for `first` and `last` that overwrite our previous Ruby versions of those methods.

The Java 8 case was a direct conflict with an actual Java method `map`. In that case, we agree that we should not make any Java methods inaccessible via our custom Ruby code. But I think that same logic applies when aliasing Java methods: in the process of defining Ruby-like aliases we should not overwrite Ruby behavior we added with the same name.

So priorities when defining methods in our Java proxy classes:

* Actual Java names should always call those Java methods (this covers the Java 8 `map` situation).
* Ruby helper methods should always call the helpers, provided they do not mask a Java method name.
* Aliases added for Java names should not overwrite either of these two.

The case I see being problematic for users like @jimtng is when you have a heterogeneous set of Ruby and Java collections and want to use the same code to work with both. The `ruby_first` method won't exist on the Ruby collections and the `first` method doesn't do what you expect on the Java collections.

@jimtng While we debate the path forward, your best workaround would be to manually fix those methods as I described above: https://github.com/jruby/jruby/issues/8526#issuecomment-2593462082

**Comment by kares** [comment]
I believe we will open a can of worms here, if this gets resolved but I raised my concern and that's about how far I can go.

**Comment by jimtng** [comment]
This issue only affected one point in our code and we've changed it to #ruby_first. I'm happy to close this isssue for me.

**Comment by headius** [comment]
Fair enough! I wish there were a clear answer going forward.
